opensearch/src/main/java/org/opensearch/sql/opensearch/data/type/OpenSearchDataType.java

```diff
@@ -6,82 +6,220 @@
+    }
+    return this;
+  }
 
   /**
-   * Parent of current type.
+   * A constructor function which builds proper `OpenSearchDataType` for given mapping `Type`.
+   * @param mappingType A mapping type.
+   * @return An instance or inheritor of `OpenSearchDataType`.
    */
-  private final List<ExprType> parents;
+  public static OpenSearchDataType of(MappingType mappingType) {
+    ExprCoreType exprCoreType = ExprCoreType.UNKNOWN;
+    switch (mappingType) {
+      // TODO update these 2 below #1038 https://github.com/opensearch-project/sql/issues/1038
+      case Text: return new OpenSearchTextType();
+      case Keyword: exprCoreType = ExprCoreType.STRING;
+        break;
+      case Byte: exprCoreType = ExprCoreType.BYTE;
+        break;
+      case Short: exprCoreType = ExprCoreType.SHORT;
+        break;
+      case Integer: exprCoreType = ExprCoreType.INTEGER;
+        break;
+      case Long: exprCoreType = ExprCoreType.LONG;
+        break;
+      case HalfFloat: exprCoreType = ExprCoreType.FLOAT;
+        break;
+      case Float: exprCoreType = ExprCoreType.FLOAT;
+        break;
+      case ScaledFloat: exprCoreType = ExprCoreType.DOUBLE;
+        break;
+      case Double: exprCoreType = ExprCoreType.DOUBLE;
+        break;
+      case Boolean: exprCoreType = ExprCoreType.BOOLEAN;
+        break;
+      // TODO: check formats, it could allow TIME or DATE only
+      case Date: exprCoreType = ExprCoreType.TIMESTAMP;
+        break;
+      case Object: exprCoreType = ExprCoreType.STRUCT;
+        break;
+      case Nested: exprCoreType = ExprCoreType.ARRAY;
+        break;
+      case GeoPoint: return new OpenSearchGeoPointType();
+      case Binary: return new OpenSearchBinaryType();
+      case Ip: return new OpenSearchIpType();
+      default:
+        throw new IllegalArgumentException(mappingType.toString());
+    }
+    var res = new OpenSearchDataType(mappingType);
+    res.exprCoreType = exprCoreType;
+    return res;
+  }
+
+  protected OpenSearchDataType(MappingType mappingType) {
+    this.mappingType = mappingType;
+  }
+
   /**
-   * JDBC type name.
+   * A constructor function which builds proper `OpenSearchDataType` for given {@link ExprType}.
+   * @param type A type.
+   * @return An instance of `OpenSearchDataType`.
    */
-  private final String jdbcType;
+  public static OpenSearchDataType of(ExprType type) {
+    if (type instanceof OpenSearchDataType) {
+      return (OpenSearchDataType) type;
+    }
+    return new OpenSearchDataType((ExprCoreType) type);
+  }
 
-  @Override
-  public List<ExprType> getParent() {
-    return parents;
+  protected OpenSearchDataType(ExprCoreType type) {
+    this.exprCoreType = type;
   }
 
+  protected OpenSearchDataType() {
+  }
+
+  // object and nested types have properties - info about embedded types
+  @Getter
+  @EqualsAndHashCode.Exclude
+  Map<String, OpenSearchDataType> properties = new LinkedHashMap<>();
+
+  // text could have fields
+  @Getter
```

Comment:
fields and properties can exist on the same type.

See [this](https://stackoverflow.com/questions/52605536/what-is-the-difference-between-a-field-and-a-property-in-elasticsearch) question, especially [this](https://stackoverflow.com/questions/52605536/what-is-the-difference-between-a-field-and-a-property-in-elasticsearch#comment106056499_52607059) comment.

We'd need to understand how they can interact together.

Reply:
I need examples of mapping and data to play with. I asked community for examples, but there is no response yet.

Current implementation can parse `fields` and `properties` both, but it maybe not used properly.